### PR TITLE
Implement new config spec #449

### LINKED ISSUE #449: Implement new config spec
changes required to helix-cli based from discussion in https://github.com/adobe/helix-shared/pull/12:

- [x] `hlx up` uses a built-in default config and issues a warning if no config is present
- [x] `hlx up` provides an option to simulate a domain with `--domain` (moved to #108)
- [x] `hlx deploy` fails if no config is present (with an informative error)
- [x] `hlx deploy` lists all strains affected
- [x] `hlx deploy` suggests new strains if non affected
- [x] `hlx deploy` adds missing stain with `--add` (moved to #456)
- [x] `hlx deploy` uses _new_ action name format
- [x] `hlx deploy` writes correct effective config based on resolved actions (moved to #457)
- [x] `hlx publish` can deal with new code format
- [x] `hlx perf` can deal with config
- [x] `hlx deploy` updates the `package` property and writes back the `helix-config.yaml`
- [x] `hlx publish` updates the strains that contain a `package` property
- [x] update demos with correct config

### LINKED ISSUE #456: [hlx deploy] add missing strain with --add
from #449 :

In order to increase the visibility of changes happening during deployment, `hlx deploy` will list all strain names that will be affected by the deployment.

- `hlx deploy` gets the current git-remote `git remote get-url origin` as `$CURRENT_CODE_REPO` and checks all strains for a `code` property that matches the `$CURRENT_CODE_REPO`.
- If no strains are affected, `hlx deploy` will print a new strain config to `stdout` that points to the new `code` location (`$CURRENT_CODE_REPO`), copies all other values from `default` except for `url` or `condition`. The new strain will have an auto-generated, hard-to-guess name, so that it cannot unwittingly be accessed.

## new feature

- When running `hlx deploy --add=foo` the new strain will be added to the configuration file automatically and `hlx deploy` will instead show instructions on accessing the strain.
- When running `hlx deploy --add=default` the default strain will be created or updated.

### PR DESCRIPTION
**Pull request by tripodsan**
fixes #449, fixes #456

- Implement new config spec #449
- [hlx deploy] add missing strain with --add #456

### CODE REVIEW

File: src/up.cmd.js [hunk 1]
```diff
@@ -92,6 +100,11 @@ class UpCommand extends BuildCommand {
   async run() {
     await super.init();
 
+    if (!await this.config.hasFile()) {
+      this.log.warn(chalk`No {cyan helix-config.yaml}. Please add one before deployment.`);
```
[the remark this answers is not in the document]
Comment (trieloff) [reply src/up.cmd.js]: I think a warning is appropriate. Maybe we can be even more helpful in the message and show what a good starter config might look like (i.e. "copy `effective-helix-config,yaml` to get started")

[the remark this answers is not in the document]
Comment (rofe) [reply src/up.cmd.js]: Then we should rethink our first run experience. I just think it's weird to do `hlx demo` and get a warning thrown my way at the very next step...

**Comment by rofe** [comment]
> `hlx deploy` will fail w/o a config.

Yes, that's why I think `deploy` is the best place to suggest adding a config. At this stage, the user is done playing around with `demo` and `up` and wants to get more serious. So `hlx` should give guidance how to add a proper config here.

File: src/up.cmd.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (tripodsan) [reply src/up.cmd.js]: I think we should only give a warning, once an origin is defined. so the experience is a bit better.
so:
- hlx demo
- hlx up (no warning)
- git remote add ....
- hlx up (warning)

**Comment by trieloff** [comment]
@tripodsan I'm planning to start work on #401 and will turn this into a PR to your PR.

**Comment by rofe** [comment]
The same error message is displayed in a full demo on `up`. I guess the `helix-config.yaml` hasn't been updated yet.

**Comment by tripodsan** [comment]
> Which configuration? What data? There isn't even a `helix-config.yaml` in the project yet. What is the user expected to do? I think the error should either explain the next steps, or point to a doc URL where the user can learn more.

hmm.. the error needs to be better worded. you need a config in order to be able to deploy.

**Comment by tripodsan** [comment]
> I guess the helix-config.yaml hasn't been updated yet. But why can't a user run up without a strains config?

yes. I forgot to update the demos. you can run `up` without a config. but not with a invalid one :-)

**Pull request merged**